---
title: "Sequoia Dragen VSP Analysis"
author: "Pender, J."
date: "Last updated on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: default
  fig_caption: yes
  pdf_document: null
editor_options:
  chunk_output_fraction: console
---


# Setting up working directory, packages
```{r}
setwd("~/github/sequoia-dragen/outputs-sequoia/vsp")
rm(list = ls())

library(readr)
library(jsonlite)
library(dplyr)
library(tidyr)
library(ggplot2)
library(purrr)
library(patchwork)
library(tidyverse)
library(stringr)
library(glue)
library(writexl)
library(readxl)
library(openxlsx)
library(ggpubr)
library(forcats)
library(grid)
# library(ComplexUpset)

```

# Load files
```{r}

kraken_files <- list.files("../../inputs-sequoia/kraken/", pattern = "*.krk.txt", full.names = TRUE)
df_kraken <- data.frame()

panel <- read_excel("../../inputs-sequoia/genus_family.xlsx")

dragen_report <- read_excel("../../inputs-sequoia/dragen_report.xlsx")

```

# Optional: single kraken file option
```{r}
kraken_df <- read.table("../../inputs-sequoia/kraken/1130_ESP_SOL_S235_L002.krk.txt", sep = "\t", header = FALSE, stringsAsFactors = FALSE)

kraken_df <- read_tsv(
  "../../inputs-sequoia/kraken/1130_ESP_SOL_S235_L002.krk.txt",
  col_names = c(
    "percentage",
    "reads_clade",
    "reads_taxon",
    "rank_code",
    "taxon_ID",
    "taxon_name"
  )
  
)
```

# Read in Kraken data (for PMMoV and other non-targets, etc)
```{r}
kraken_files <- list.files("../../inputs-sequoia/kraken/", pattern = "\\.krk\\.txt$", full.names = TRUE)

# Read and combine all files
kraken_df <- map_dfr(
  kraken_files,
  function(file) {
    read_tsv(
      file,
      col_names = c(
        "percentage",
        "reads_clade",
        "reads_taxon",
        "rank_code",
        "taxon_ID",
        "taxon_name"
      ),
      show_col_types = FALSE
    ) %>%
    mutate(sample = sub("^(.*)_.*_.*$", "\\1", sub("\\..*", "", basename(file)))
  )
  }
)

# Move 'sample' to be the first column
kraken_df <- kraken_df %>% select(sample, everything())

# Format sample name to be consistent with dragen format
kraken_df$sample <- gsub("_", "-", kraken_df$sample)

# Filter kraken_df for taxon_ID == 12239 and keep only necessary columns
kraken_pmmov <- kraken_df %>%
  filter(taxon_ID == 12239) %>%
  select(sample, percent.pmmov.kraken = percentage, pmmov.reads.kraken = reads_taxon)

```

# Optional: single Dragen .json file option
```{r}
# load DRAGEN output
file <- "../../inputs-sequoia/sequoia-vsp/0818-WOD-SOL.VSPv2.report.json"
json_data <- fromJSON(file, simplifyDataFrame = FALSE)

df_troubleshoot <- data.frame()
microorganisms <- json_data$targetReport$microorganisms

sample_name = sub("\\..*", "", basename(file))
df <- extract_microorganism_df(microorganisms, sample_name)

df$genus.name <- panel$Genus_Name[match(df$name, panel$Reporting_Name)]
df$strand.type <- panel$Type[match(df$name, panel$Reporting_Name)]
  
df$fraction <- case_when(
      grepl("(^|-)INF$", df.amr.markers.single$sample) ~ "Influent",
      grepl("(^|-)SOL$", df.amr.markers.single$sample) ~ "Solids",
      TRUE ~ "Other" 
)

df.removed.spike <- df %>% filter(!Name %in% bcov.spike)

df.microorganisms.single$relabund.rpkm <- as.numeric(df.microorganisms.single$rpkm) / sum(as.numeric(df.microorganisms.single$rpkm), na.rm = TRUE)

```


# Make dataframes
```{r}
json_files_vsp <- list.files("../../inputs-sequoia/sequoia-vsp/", pattern = "*.json", full.names = TRUE)
df_vsp <- data.frame()

# function to extract a df from each sample file
extract_microorganism_df <- function(microorganisms, sample_name) {
  # Always start with an NA row
  df_na <- data.frame(
    sample = sample_name,
    name = NA,
    alignedReadCount = NA,
    ani = NA,
    coverage = NA,
    medianDepth = NA,
    rpkm = NA,
    stringsAsFactors = FALSE
  )
  
  # If there are microorganisms, generate rows for them
  if (length(microorganisms) > 0) {
    df_data <- do.call(rbind, lapply(microorganisms, function(micro) {
      data.frame(
        sample = sample_name,
        name = micro$name,
        alignedReadCount = micro$alignedReadCount,
        ani = micro$ani,
        coverage = micro$coverage,
        medianDepth = micro$medianDepth,
        rpkm = micro$rpkm,
        stringsAsFactors = FALSE
      )
    }))
    df <- rbind(df_na, df_data)
  } else {
    df <- df_na
  }
  
  return(df)
}


  df_vsp <- data.frame()
  
  for (file in json_files_vsp) {
    # Load JSON
    json_data <- fromJSON(file, simplifyDataFrame = FALSE)
    microorganisms <- json_data$targetReport$microorganisms
    sample_name <- sub("\\..*", "", basename(file))
    
    # Extract per-sample microorganism data
    df <- extract_microorganism_df(microorganisms, sample_name)
    
    # Append to main dataframe
    df_vsp <- rbind(df_vsp, df)
  }

```

#Optional: Check samples with no or one virus
```{r}
# Shows NA or single-virus samples
single_entry_df <- df_vsp %>%
  group_by(sample) %>%
  filter(n() == 1) %>%
  ungroup()

print(single_entry_df)

# Shows NA or single-virus samples
dual_entry_df <- df_vsp %>%
  group_by(sample) %>%
  filter(n() == 2) %>%
  ungroup()

print(dual_entry_df)

single_entry_df_genera <- df_vsp_genera %>%
  group_by(sample) %>%
  filter(n() == 1) %>%
  ungroup()

print(single_entry_df_genera)

```

# List of sample partitions and spike definitions
```{r}
controls <- c("BLANK-1", "BLANK-2", "VIRCELL-ADV41", "VIRCELL-H1", "ZEPTO-ADV1", "ZEPTO-H1")
nonsamples <- c("AACTTATCCT-ACGCCAGTAC", "CCAATGATAC-CGGCACACTC", "CCATGGTATA-CTCCGCGAGA",
                      "CGATCTGTGA-AGTCTTCCTA", "GTAACAATCT-GTATACAGAG", "GTAGGCGAGC-TGCGCTCCTA",
                      "GTGGACAAGT-AAGGCCACGG", "TCTCAATACC-CGCGAACGTC", "TCTTGTCGGC-ACACAGGTGG",
                      "TGCAAGATAA-CCGTCCTCAA", "TTAGACCATG-CTTCCTAGGA",
                      "Undetermined-lane2", "Undetermined-lane3"
                      )
tri.spike.samples  <- df_vsp$sample[grepl("^0718-WOD-INF", df_vsp$sample)]


bcov.spike <- "Human coronavirus OC43 (HCoV_OC43)"

tri.spike <- c("Human coronavirus OC43 (HCoV_OC43)", 
               "Human adenovirus F", 
               "Influenza A virus (H1N1)")

```

# Expand data frame to include info on everything on the panel for each sample
```{r}
# creates grid of all unique samples and all viruses on the panel for each sample
df_vsp_expanded <- expand_grid(
  sample = unique(df_vsp$sample),
  name = unique(panel$Reporting_Name)
)

# # old version - All virus hits in dataset get represented per sample, not all in panel
# df_vsp_expanded <- expand_grid(
#   sample = unique(df_vsp$sample),
#   name = unique(df_vsp$name)
# )

# Add columns
# If a virus was present in a sample, the columns will have values, if absent, NA
df_vsp_expanded2 <- df_vsp_expanded %>%
  left_join(df_vsp, by = c("sample", "name"))

# Get rid of placeholder NA row from earlier for samples that didn't have hits
df_vsp_expanded3 <- df_vsp_expanded2 %>% filter(!is.na(name))

df_vsp_expanded4 <- df_vsp_expanded3 %>%
  mutate(
    alignedReadCount = replace_na(alignedReadCount, 0),
    rpkm = replace_na(rpkm, 0)
  )

df_vsp_expanded4 <- df_vsp_expanded4 %>%
  left_join(panel %>% select(Reporting_Name, Genus_Name, strand.type, taxid),
            by = c("name" = "Reporting_Name"))


df_vsp_expanded4$fraction <- case_when(
      grepl("(^|-)INF$", df_vsp_expanded4$sample) ~ "Influent",
      grepl("(^|-)SOL$", df_vsp_expanded4$sample) ~ "Solids",
      TRUE ~ "Other"
    )

# Take out fraction from the sample name (0505-MOD-SOL -> 0505-MOD) and replace UCD with COD for solids samples
df_vsp_expanded4$sample.edit <- gsub("-(SOL|INF)", "", df_vsp_expanded4$sample)
df_vsp_expanded4$sample.edit[df_vsp_expanded4$fraction == "Solids"] <- gsub("UCD", "COD", df_vsp_expanded4$sample.edit[df_vsp_expanded4$fraction == "Solids"])

df_vsp_expanded5 <- df_vsp_expanded4 %>%
  filter(fraction %in% c("Solids", "Influent", "Other")) %>%
  mutate(fraction = factor(fraction, levels = c("Influent", "Solids", "Other")))

# Add label for what kind of sample/control/etc it is
df_vsp_expanded5 <- df_vsp_expanded5 %>%
  mutate(
    type = case_when(
      sample %in% controls ~ "controls",
      sample %in% nonsamples ~ "nonsamples",
      sample %in% tri.spike.samples ~ "tri-spike samples",
      TRUE ~ "single-spike samples" # if none of these, then single-spike is default
    )
  )

# adds location from sample name, capturing everything after the first '-' and up to the next '-' or end of string
df_vsp_expanded5$WWTP.abbr <- ifelse(
  df_vsp_expanded5$type %in% c("tri-spike samples", "single-spike samples"),
  sub("^[^\\-]+-([^\\-]+).*", "\\1", df_vsp_expanded5$sample.edit),
  NA
)

df_vsp_expanded5 <- df_vsp_expanded5 %>%
  mutate(
    WWTP = if_else(
      type %in% c("tri-spike samples", "single-spike samples"),
      case_when(
        WWTP.abbr == "COD" ~ "City of Davis",
        WWTP.abbr == "UCD" ~ "UC Davis",
        WWTP.abbr == "ESP" ~ "Esparto",
        WWTP.abbr == "LOB" ~ "Los Banos",
        WWTP.abbr == "TUR" ~ "Turlock",
        WWTP.abbr == "WIN" ~ "Winters",
        WWTP.abbr == "WOD" ~ "Woodland",
        WWTP.abbr == "MER" ~ "Merced",
        WWTP.abbr == "MOD" ~ "Modesto",
        TRUE ~ NA_character_
      ),
      NA_character_
    )
  )

df_vsp_expanded5$city <- ifelse(
  df_vsp_expanded5$WWTP %in% c("City of Davis", "UC Davis"),
  "Davis",
  df_vsp_expanded5$WWTP
)

df_vsp_expanded5 <- df_vsp_expanded5 %>%
  mutate(WWTP_fraction = paste(WWTP, fraction, sep = " - "))

levels.fraction <- c("Influent", "Solids")
levels.WWTP <- c("City of Davis", "UC Davis", "Esparto", "Los Banos", "Turlock", "Winters", "Woodland", "Merced", "Modesto")
levels.city <- c("Davis", "Esparto", "Los Banos", "Turlock", "Winters", "Woodland", "Merced", "Modesto")         
levels.WWTP_fraction <- c("UC Davis - Influent", "City of Davis - Solids",
                          "Esparto - Influent", "Esparto - Solids", 
                          "Los Banos - Influent", "Los Banos - Solids",
                          "Turlock - Influent", "Turlock - Solids", 
                          "Winters - Influent", "Winters - Solids", 
                          "Woodland - Influent", "Woodland - Solids", 
                          "Merced - Solids", "Modesto - Solids")

custom.colors <- c(
  "Influent" = "#88CCEE",
  "Solids" = "#332288",
  "UC Davis" = "#8DD3C7", "City of Davis" = "#8DD3C7", "Davis" = "#8DD3C7", 
  "UC Davis - Influent" = "#8DD3C7", "City of Davis - Solids" = "#5CB8A7", 
  "Davis - Influent" = "#8DD3C7", "Davis - Solids" = "#5CB8A7",
  "Esparto" = "#FFFFB3", "Esparto - Influent" = "#FFFFB3", "Esparto - Solids" = "#F7F77D",
  "Los Banos" = "#BEBADA", "Los Banos - Influent" = "#BEBADA", "Los Banos - Solids" = "#9D97C4",
  "Turlock" = "#FB8072", "Turlock - Influent" = "#FB8072", "Turlock - Solids" = "#E95540",
  "Winters" = "#80B1D3", "Winters - Influent" = "#80B1D3", "Winters - Solids" = "#4C91C1",
  "Woodland" = "#FDB462", "Woodland - Influent" = "#FDB462", "Woodland - Solids" = "#E38B30",
  "Merced" = "#B3DE69", "Merced - Solids" = "#B3DE69",
  "Modesto" = "#FCCDE5", "Modesto - Solids" = "#FCCDE5"
)
  
df_vsp_expanded5 <- df_vsp_expanded5 %>%
  mutate(
    WWTP = factor(WWTP, levels = levels.WWTP),
    city = factor(city, levels = levels.city),
    WWTP_fraction = factor(WWTP_fraction, levels = levels.WWTP_fraction)
  )

```

# Add ddPCR data
```{r}
 # adds ddpcr columns associated to each sample
solids_ddpcr <- read_excel("../../inputs-sequoia/solids_ddpcr_eurofins.xlsx")
influent_ddpcr <- read_excel("../../inputs-sequoia/influent_sequoia_ddpcr.xlsx")

df_vsp_expanded6 <- df_vsp_expanded5 %>%
  mutate(
    BCoV.ddpcr.recovery = case_when(
      fraction == "Influent" ~ as.numeric(influent_ddpcr$bCoVRecovery_frac[match(sample.edit, influent_ddpcr$sample.edit)]),
      fraction == "Solids" ~ as.numeric(solids_ddpcr$BCoV_Recovery[match(sample.edit, solids_ddpcr$sample.edit)]),
      TRUE ~ NA_real_
    ),
    PMMoV.ddpcr.conc = case_when(
      fraction == "Influent" ~ as.numeric(influent_ddpcr$PV_Concentration_Merged[match(sample.edit, influent_ddpcr$sample.edit)]),
      fraction == "Solids" ~ as.numeric(solids_ddpcr$PMMoV_gc_per_g_dry_weight[match(sample.edit, solids_ddpcr$sample.edit)]),
      TRUE ~ NA_real_
    ),
    BCoV.ddpcr.conc = if_else(
      fraction == "Influent",
      as.numeric(influent_ddpcr$BC_Concentration_Merged[match(sample.edit, influent_ddpcr$sample.edit)]),
      NA_real_
    ),
    frac.solid = if_else(
      fraction == "Solids",
      as.numeric(solids_ddpcr$Frac_Solid[match(sample.edit, solids_ddpcr$sample.edit)]),
      NA_real_
    )
  ) 

# Extract BCoV RPKM per sample
bcov_rpkm_lookup <- df_vsp_expanded6 %>%
  filter(name == bcov.spike) %>%
  select(sample, bcov_rpkm = rpkm)

# Join and compute scaling factor and normalized RPKM
df_vsp_expanded7 <- df_vsp_expanded6 %>%
  left_join(bcov_rpkm_lookup, by = "sample") %>%
  mutate(
    BCoV.recovery.scale.factor = if_else(
      !is.na(bcov_rpkm) & bcov_rpkm != 0,
      BCoV.ddpcr.recovery / bcov_rpkm,
      NA_real_
    ),
    rpkm.norm.to.BCoV.recovery = if_else(
      !is.na(BCoV.recovery.scale.factor),
      rpkm * BCoV.recovery.scale.factor, # multiply each virus's rpkm by the scaling factor
      NA_real_
    )
  )

df_vsp_expanded7$rpkm.norm.to.BCoV.rpkm <- df_vsp_expanded7$rpkm / df_vsp_expanded7$bcov_rpkm


# Join Kraken data
df_vsp_expanded7 <- df_vsp_expanded7 %>%
  left_join(kraken_pmmov, by = "sample")

df_vsp_expanded_complete <- df_vsp_expanded7
  
```

# segregate samples
```{r}
df_controls <- df_vsp_expanded_complete %>%
  filter(type == "controls")

df_nonsamples <- df_vsp_expanded_complete %>%
  filter(type == "nonsamples")

df_tri_spike_samples <- df_vsp_expanded_complete %>%
  filter(type == "tri-spike samples")

df_single_spike_samples <- df_vsp_expanded_complete %>%
  filter(type == "single-spike samples")

# Create dataframe of samples from the main block that do not have BCoV so that we can remove these
df_no_OC43 <- df_single_spike_samples %>% # Gets all sample IDs where OC43 has alignedReadCount == 0.
  filter(
    sample %in% (
      df_single_spike_samples %>%
        filter(name == bcov.spike, alignedReadCount == 0) %>%
        pull(sample) %>%
        unique()
    )) %>%  # Keeps all hit rows for samples that didn't have OC43
  filter(
    (name == bcov.spike & alignedReadCount == 0) |
    (alignedReadCount > 0)
  )

# Takes out rows that do not have the BCoV spike-in
df_single_spike_samples_has_OC43 <- df_single_spike_samples %>%
  filter(!sample %in% unique(df_no_OC43$sample))
  
# Grabs samples names that end with INF or SOL. Does not include 0718-WOD-INF- replicates
df_influent_has_OC43 <- df_single_spike_samples_has_OC43 %>% filter(fraction == "Influent")
df_solids_has_OC43 <- df_single_spike_samples_has_OC43 %>% filter(fraction == "Solids")

# Create subsets that include only WWTPs that have both solids and influent samples
direct.compare_list <- c("Esparto", "Los Banos", "Turlock", "Winters", "Woodland")

df_single_spike_samples_has_OC43_direct.compare <- df_single_spike_samples_has_OC43 %>%
  filter(city %in% direct.compare_list)

df_influent_has_OC43_direct.compare <- df_influent_has_OC43 %>%
  filter(city %in% direct.compare_list)

df_solids_has_OC43_direct.compare <- df_solids_has_OC43 %>%
  filter(city %in% direct.compare_list)

```

# Optional: Export unique sample info to Excel
```{r}
# Subset for unique samples with relevant metadata
df_single_spike_samples_export <- df_single_spike_samples_has_OC43 %>%
  distinct(sample, .keep_all = TRUE) %>%
  select(sample, fraction, WWTP, city, WWTP_fraction)

df_influent_export <- df_influent_has_OC43 %>%
  distinct(sample, .keep_all = TRUE) %>%
  select(sample, fraction, WWTP, city, WWTP_fraction)

df_solids_export <- df_solids_has_OC43 %>%
  distinct(sample, .keep_all = TRUE) %>%
  select(sample, fraction, WWTP, city, WWTP_fraction)

# Export to xlsx
write_xlsx(df_single_spike_samples_export, "single_spike_samples_filtered_list.xlsx")
write_xlsx(df_influent_export, "influent_samples_filtered_list.xlsx")
write_xlsx(df_solids_export, "solids_samples_filtered_list.xlsx")
```

# Remove spikes
```{r}
remove_spikes <- function(df, spike_names) {
  df %>% filter(!name %in% spike_names)
}

df_tri_spike_samples_spike.removed <- remove_spikes(df_tri_spike_samples, tri.spike)
df_samples_spike.removed <- remove_spikes(df_single_spike_samples_has_OC43, bcov.spike)
df_influent_spike.removed <- remove_spikes(df_influent_has_OC43, bcov.spike)
df_solids_spike.removed <- remove_spikes(df_solids_has_OC43, bcov.spike)

df_samples_spike.removed_direct.compare <- remove_spikes(df_single_spike_samples_has_OC43_direct.compare, bcov.spike)
df_influent_spike.removed_direct.compare <- remove_spikes(df_influent_has_OC43_direct.compare, bcov.spike)
df_solids_spike.removed_direct.compare <- remove_spikes(df_solids_has_OC43_direct.compare, bcov.spike)

```

# Optional: Consolidate to Genera
```{r}
summarize_by_genus <- function(df) {
  df %>%
    group_by(sample, genus.name) %>%
    summarize(
      alignedReadCount = sum(alignedReadCount, na.rm = TRUE),
      rpkm = sum(rpkm, na.rm = TRUE),
      .groups = "drop"
    )
}

df_vsp_genera <- summarize_by_genus(df_vsp_expanded_complete) # all samples including controls

# nonsamples
df_vsp_controls_genera <- summarize_by_genus(df_controls)
df_vsp_nonsamples_genera <- summarize_by_genus(df_nonsamples)

# samples without spikes removed
df_tri_spike_samples_genera <- summarize_by_genus(df_tri_spike_samples)
df_samples_genera <- summarize_by_genus(df_single_spike_samples_has_OC43)
df_influent_genera <- summarize_by_genus(df_influent_has_OC43)
df_solids_genera <- summarize_by_genus(df_solids_has_OC43)

# samples with spikes removed
df_tri_spike_samples_spike.removed_genera <- summarize_by_genus(df_tri_spike_samples_spike.removed)
df_samples_spike.removed_genera <- summarize_by_genus(df_samples_spike.removed)
df_influent_spike.removed_genera <- summarize_by_genus(df_influent_spike.removed)
df_solids_spike.removed_genera <- summarize_by_genus(df_solids_spike.removed)



```

# Optional: Export json genus level
```{r}
# NOTE CHANGE THIS to export to many json files, not just one
export_genus_abundance_to_json <- function(df, output_folder, file) {
  # Convert to a named vector: names = Genus_Name, values = Abundance
  aggregated_vector <- setNames(df$rpkm, df$genus.name)
  aggregated_list   <- as.list(aggregated_vector)

  # Create the output filename by stripping everything after the first dot
  base_name <- sub("\\..*", "", basename(file))
  out_filename <- paste0(base_name, ".json")
  
  # Construct full output folder path and create it if it doesn't exist
  output_folder <- file.path("genera_json_files", subfolder)
  dir.create(output_folder, showWarnings = FALSE, recursive = TRUE)
  
  # Full path for the output file
  full_out_path <- file.path(output_folder, out_filename)

  # Write JSON to file
  write_json(aggregated_list, full_out_path, pretty = TRUE, auto_unbox = TRUE)

  return(full_out_path)
}

```

# Add relative abundances and drop viruses that don't show up
```{r}
process2 <- function(df, drop.absent.viruses = TRUE) {
  df <- df %>%
    group_by(sample) %>%
    mutate(
      relabund.rpkm = if (sum(rpkm) == 0) 0 else rpkm / sum(rpkm),
      percent.rpkm = round(relabund.rpkm * 100, 2)
    ) %>%
    ungroup()
  
  # Arrange viruses in descending order. Take out desc if want ascending
  df$name <- factor(df$name, levels = df %>%
                           group_by(name) %>%
                           summarize(total.relabund.rpkm = sum(relabund.rpkm)) %>%
                           arrange(desc(total.relabund.rpkm)) %>%
                           pull(name))
  
  if(drop.absent.viruses){
    # Gets rid of virus names expanded to every sample that do not show up across the df
    df <- df %>%
    anti_join(
      df %>%
        group_by(name) %>%
        summarize(total_reads = sum(alignedReadCount, na.rm = TRUE)) %>%
        filter(total_reads == 0),
      by = "name"
    )
  }
  return(df)
}

df_vsp_expanded_complete_2 <- process2(df_vsp_expanded_complete)

# nonsamples
df_controls_2 <- process2(df_controls)
df_nonsamples_2 <- process2(df_nonsamples)

# samples without spikes removed
df_tri_spike_samples_2 <- process2(df_tri_spike_samples)
df_single_spike_samples_has_OC43_2 <- process2(df_single_spike_samples_has_OC43)
df_influent_has_OC43_2 <- process2(df_influent_has_OC43)
df_solids_has_OC43_2 <- process2(df_solids_has_OC43)

# samples with spikes removed
df_tri_spike_samples_spike.removed_2 <- process2(df_tri_spike_samples_spike.removed)
df_samples_spike.removed_2 <- process2(df_samples_spike.removed)
df_influent_spike.removed_2 <- process2(df_influent_spike.removed)
df_solids_spike.removed_2 <- process2(df_solids_spike.removed)


# direct compare - 5 WWTPs - without spike removed
df_single_spike_samples_has_OC43_direct.compare_2 <- process2(df_single_spike_samples_has_OC43_direct.compare)
df_influent_has_OC43_direct.compare_2 <- process2(df_influent_has_OC43_direct.compare)
df_solids_has_OC43_direct.compare_2 <- process2(df_solids_has_OC43_direct.compare)

# direct compare - 5 WWTPs - spike removed
df_samples_spike.removed_direct.compare_2 <- process2(df_samples_spike.removed_direct.compare)
df_influent_spike.removed_direct.compare_2 <- process2(df_influent_spike.removed_direct.compare)
df_solids_spike.removed_direct.compare_2 <- process2(df_solids_spike.removed_direct.compare)


# with all viruses retained including ones that have no hits in df
# samples without spikes removed, all viruses retained
df_single_spike_samples_has_OC43_2_undropped <- process2(df_single_spike_samples_has_OC43, drop.absent.viruses = F)
df_influent_has_OC43_2_undropped <- process2(df_influent_has_OC43, drop.absent.viruses = F)
df_solids_has_OC43_2_undropped <- process2(df_solids_has_OC43, drop.absent.viruses = F)

# samples with spikes removed
df_samples_spike.removed_2_undropped <- process2(df_samples_spike.removed, drop.absent.viruses = F)
df_influent_spike.removed_2_undropped <- process2(df_influent_spike.removed, drop.absent.viruses = F)
df_solids_spike.removed_2_undropped <- process2(df_solids_spike.removed, drop.absent.viruses = F)

```

# Number of samples and viable samples per location
```{r}
summarize_samples <- function(df){
  df <- df %>%
    group_by(fraction, WWTP) %>%
    summarise(unique_samples = n_distinct(sample)) %>%
    arrange(fraction, WWTP)
  return(df)
}

df_all_sample_summary <- summarize_samples(df_vsp_expanded_complete)
df_all_viable_sample_summary <- summarize_samples(df_single_spike_samples_has_OC43)

```

# BCoVscatterplot
```{r}
df_bcov_samples <- df_vsp_expanded_complete_2 %>%
  filter(name == bcov.spike)

df_bcov_influent <- df_influent_has_OC43_direct.compare_2 %>%
  filter(name == bcov.spike)

df_bcov_solids <- df_solids_has_OC43_direct.compare_2 %>%
  filter(name == bcov.spike)

df_bcov_solids_filtered <- df_bcov_solids %>%
  filter(bcov_rpkm < 1000)

ggplot(df_bcov_solids, aes(x = sample, y = bcov_rpkm)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "red") +
  stat_regline_equation(
    aes(label = ..eq.label..), 
    label.x = 2, label.y = .005
  ) +
  stat_cor(
    aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
    label.x = 2, label.y = .004
  ) +
  theme_minimal()


```

# Coverage Plots
```{r}
plot_metrics <- function(df, output_prefix = "output") {
  # Ensure Name is a factor
  df$name <- factor(df$name)
  
  # Group positions for vertical lines
  group_positions <- seq(1.5, length(levels(df$name)) - 0.5, by = 1)
  
  # Convert necessary columns to numeric
  cols_to_convert <- c("relabund.rpkm", "ani", "coverage", "rpkm")
  df[cols_to_convert] <- lapply(df[cols_to_convert], function(col) as.numeric(as.character(col)))
  
  # Define a helper function for boxplots
  make_boxplot <- function(data, yvar, ylab_text, ybreaks = NULL, ylim_vals = NULL, clean_x = FALSE) {
    p <- ggplot(data, aes(x = name, y = .data[[yvar]])) +
      xlab("Virus") +
      ylab(ylab_text) +
      geom_vline(xintercept = group_positions, linefraction = "dotted", color = "black") +
      stat_boxplot(geom = "errorbar") +
      geom_boxplot(fill = "gray") +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        text = element_text(size = 11.85),
        panel.border = element_rect(color = "black", fill = NA, size = 1),
        panel.background = element_rect(fill = "white"),
        plot.margin = unit(c(0.5, 0.5, 0.5, 2), "cm")
      )
    
    if (!is.null(ybreaks)) {
      p <- p + scale_y_continuous(breaks = ybreaks)
    }
    if (!is.null(ylim_vals)) {
      p <- p + coord_cartesian(ylim = ylim_vals)
    }
    if (clean_x) {
      p <- p + theme(
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()
      )
    }
    return(p)
  }
  
  # Create plots
  abundance_plot <- make_boxplot(df, "rpkm", "Abundance (RPKM)")
  coverage_plot  <- make_boxplot(df, "coverage", "Coverage", seq(0, 1, 0.1), c(0, 1))
  ani_plot       <- make_boxplot(df, "ani", "Average Nucleotide Identity (ANI)", seq(0.6, 1, 0.1), c(0.6, 1))
  relabund_plot  <- make_boxplot(df, "relabund.rpkm", "Relative Abundance (RPKM)", seq(0, 1, 0.1), c(0, 1))

  # Clean versions for combining
  abundance_plot_clean <- make_boxplot(df, "rpkm", "Abundance (RPKM)", clean_x = TRUE)
  coverage_plot_clean  <- make_boxplot(df, "coverage", "Coverage", seq(0, 1, 0.1), c(0, 1), clean_x = TRUE)
  ani_plot_clean       <- make_boxplot(df, "ani", "Average Nucleotide Identity (ANI)", seq(0.6, 1, 0.1), c(0.6, 1), clean_x = TRUE)
  relabund_plot_clean  <- make_boxplot(df, "relabund.rpkm", "Relative Abundance (RPKM)", seq(0, 1, 0.1), c(0, 1), clean_x = TRUE)
  
  # Save individual plots
  ggsave(paste0(output_prefix, "_abundance.pdf"), abundance_plot, width = 20, height = 7)
  ggsave(paste0(output_prefix, "_coverage.pdf"), coverage_plot, width = 20, height = 7)
  ggsave(paste0(output_prefix, "_ani.pdf"), ani_plot, width = 20, height = 7)
  ggsave(paste0(output_prefix, "_relabund.pdf"), relabund_plot, width = 20, height = 7)
  
  # Combine and save combined plot
  # combined_plot <- (coverage_plot_clean / ani_plot_clean / relabund_plot) +
  combined_plot <- (coverage_plot_clean / ani_plot_clean / relabund_plot_clean / abundance_plot) +
  plot_layout(heights = c(1, 1, 1, 1)) +
  plot_annotation(
    theme = theme(
      plot.margin = unit(c(0.5, 0.5, 0.5, 2), "cm"),
      text = element_text(size = 11.85)
    )
  ) &
  labs(x = "Virus")
  
  print(combined_plot)
  ggsave(paste0(output_prefix, "_combined.pdf"), combined_plot, width = 20, height = 15)
}

# remove rows with zero aligned reads
only.hits <- function(df) {
  df %>% filter(alignedReadCount != 0)
}

df_single_spike_samples_has_OC43_only.hits <- only.hits(df_single_spike_samples_has_OC43_2)
df_samples_spike.removed_only.hits <- only.hits(df_samples_spike.removed_2)


plot_metrics(df_single_spike_samples_has_OC43_only.hits, output_prefix = "samples.with.spike")
plot_metrics(df_samples_spike.removed_only.hits, output_prefix = "samples.spike.removed")

```

# Look for cool locations
```{r}

full_grid <- expand.grid(
  name = unique(df_samples_spike.removed_processed$name),
  location = unique(df_samples_spike.removed_processed$location),
  stringsAsFactors = FALSE
)

df_virus_location_summary <- full_grid %>%
  # Count samples with alignedReadCount > 0
  left_join(
    df_samples_spike.removed_processed %>%
      filter(alignedReadCount > 0) %>%
      distinct(name, location, sample) %>%
      count(name, location, name = "sample_hits"),
    by = c("name", "location")
  ) %>%
  replace_na(list(sample_hits = 0)) %>%
  # Get total samples per location
  left_join(
    df_samples_spike.removed_processed %>%
      distinct(location, sample) %>%
      count(location, name = "total_samples"),
    by = "location"
  ) %>%
  pivot_wider(
    id_cols = name,
    names_from = location,
    values_from = c(sample_hits, total_samples),
    names_glue = "{location}_{.value}"
  ) %>%
  mutate(across(ends_with("_sample_hits"), ~replace_na(.x, 0))) %>%
  relocate(
  sort(unlist(lapply(
    sort(unique(df_samples_spike.removed_processed$location)),
    function(loc) {
      c(glue("{loc}_sample_hits"), glue("{loc}_total_samples"))
    }
  ))),
  .after = name
  ) %>%
  rowwise() %>%
  mutate(
    n_locations_found = sum(c_across(ends_with("_sample_hits")) > 0),
    total_sample_hits = sum(c_across(ends_with("_sample_hits")))
  ) %>%
  ungroup() %>%
  arrange(desc(total_sample_hits))  # <- sorting step

writexl::write_xlsx(df_virus_location_summary, "virus_location_summary.xlsx")


```

# Make spreadsheet of viruses unique to each location
```{r}
# ----------- FUNCTION START -----------
export_unique_hits_by_location <- function(df_input, use_abbr = FALSE, output_file = "unique_hits_by_location.xlsx") {
  
  # Use either location.abbr or location as the grouping variable
  df_input <- df_input %>%
    mutate(location_label = if (use_abbr) location.abbr else location)
  
  # Step 1: Summarize alignedReadCount per virus-location
  virus_location_summary <- df_input %>%
    group_by(name, location_label) %>%
    summarize(total_aligned = sum(alignedReadCount, na.rm = TRUE), .groups = "drop")
  
  # Step 2: Function to get viruses unique to a single location
  get_hits_unique_to_location <- function(location_name) {
    unique_to_location <- virus_location_summary %>%
      group_by(name) %>%
      summarize(
        location_total = sum(total_aligned[location_label == location_name]),
        other_total = sum(total_aligned[location_label != location_name]),
        .groups = "drop"
      ) %>%
      filter(location_total > 1, other_total == 0) %>%
      select(name)
    
    df_unique_hits <- df_input %>%
      filter(location_label == location_name, alignedReadCount > 1) %>%
      semi_join(unique_to_location, by = "name")
    
    return(df_unique_hits)
  }
  
  # Step 3: Apply to each location
  locations <- unique(df_input$location_label)
  locations <- as.character(locations)  # ensure character fraction
  
  unique_hits_by_location <- lapply(locations, get_hits_unique_to_location)
  names(unique_hits_by_location) <- make.unique(locations)
  
  # Step 4: Write to Excel
  wb <- createWorkbook()
  
  for (loc in names(unique_hits_by_location)) {
    df <- unique_hits_by_location[[loc]]
    addWorksheet(wb, sheetName = loc)
    writeData(wb, sheet = loc, x = df)
    setColWidths(wb, sheet = loc, cols = 1:ncol(df), widths = "auto")
  }
  
  saveWorkbook(wb, output_file, overwrite = TRUE)
}
# ----------- FUNCTION END -----------

export_unique_hits_by_location(df_samples_spike.removed_processed, 
                               use_abbr = TRUE, output_file = "unique_hits_by_location.xlsx")

export_unique_hits_by_location(df_influent_spike.removed_processed, 
                               use_abbr = FALSE, output_file = "unique_hits_by_location_influent.xlsx")

export_unique_hits_by_location(df_solids_spike.removed_processed, 
                               use_abbr = FALSE, output_file = "unique_hits_by_location_solids.xlsx")

```

# analysis
```{r}
namelist_influent <- unique(df_influent_spike.removed_2$name)
namelist_solids <- unique(df_solids_spike.removed_2$name)
namelist_influent_direct.compare <- unique(df_influent_spike.removed_direct.compare_2$name)
namelist_solids_direct.compare <- unique(df_solids_spike.removed_direct.compare_2$name)

unique_to_influent_not_in_solids <- setdiff(namelist_influent, namelist_solids)
unique_to_solids_not_in_influent <- setdiff(namelist_solids, namelist_influent)
unique_to_influent_not_in_solids_direct.compare <- setdiff(namelist_influent_direct.compare, namelist_solids_direct.compare)
unique_to_solids_not_in_influent_direct.compare <- setdiff(namelist_solids_direct.compare, namelist_influent_direct.compare)

print(unique_to_influent_not_in_solids)
print(unique_to_solids_not_in_influent)
print(unique_to_influent_not_in_solids_direct.compare)
print(unique_to_solids_not_in_influent_direct.compare)

total_rpkm_influent <- sum(df_influent_spike.removed_direct.compare_2$rpkm, na.rm = TRUE)
total_rpkm_solids <- sum(df_solids_spike.removed_direct.compare_2$rpkm, na.rm = TRUE)

total_reads_influent <- sum(df_influent_spike.removed_direct.compare_2$alignedReadCount, na.rm = TRUE)
total_reads_solids <- sum(df_solids_spike.removed_direct.compare_2$alignedReadCount, na.rm = TRUE)

total_rpkm_influent
total_rpkm_solids
total_reads_influent
total_reads_solids
```

# Consolidate samples by fraction or location
```{r}
consolidate <- function(df, consolidation.level = c("fraction", "WWTP", "WWTP_fraction")) {
  consolidation.level <- match.arg(consolidation.level)
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))

  # Add samples.total and samples.present before summarizing
  df_sample_stats <- df %>%
    group_by(.data[[consolidation.level]], name) %>%
    summarize(
      samples.total = n_distinct(sample),
      samples.present = sum(alignedReadCount > 0, na.rm = TRUE),
      .groups = "drop"
    ) %>%
    mutate(fraction.present = samples.present / samples.total)
  
  df_avg.relabund <- df %>%
  group_by(.data[[consolidation.level]], name) %>%
  summarize(relabund.rpkm.mean = mean(relabund.rpkm, na.rm = TRUE), .groups = "drop") %>%
  group_by(.data[[consolidation.level]]) %>%
  mutate( #  relabund.rpkm.mean.normalized makes the sum of viruses within each group sum to 1 to account for differences in abundance between each group
    relabund.rpkm.mean.normalized = relabund.rpkm.mean / sum(relabund.rpkm.mean, na.rm = TRUE),
    log.relabund.norm = if_else(
      relabund.rpkm.mean.normalized > 0,
      log10(relabund.rpkm.mean.normalized),
      NA_real_
    )
  ) %>%
  ungroup()

  df_avg.rpkm <- df %>%
    group_by(.data[[consolidation.level]], name) %>%
    summarize(rpkm.mean = mean(rpkm, na.rm = TRUE), .groups = "drop")
  
  df_avg.rpkm.norm.to.BCoV.recovery <- df %>%
    group_by(.data[[consolidation.level]], name) %>%
    summarize(mean.rpkm.norm.to.BCoV.recovery = mean(rpkm.norm.to.BCoV.recovery, na.rm = TRUE), .groups = "drop")
  
  df_avg.rpkm.norm.to.BCoV.rpkm <- df %>%
    group_by(.data[[consolidation.level]], name) %>%
    summarize(mean.rpkm.norm.to.BCoV.rpkm = mean(rpkm.norm.to.BCoV.rpkm, na.rm = TRUE), .groups = "drop")

  # Join the summaries
  df_combined <- df_avg.relabund %>%
  left_join(df_avg.rpkm, by = c(consolidation.level, "name")) %>%
  left_join(df_avg.rpkm.norm.to.BCoV.recovery, by = c(consolidation.level, "name")) %>%
  left_join(df_avg.rpkm.norm.to.BCoV.rpkm, by = c(consolidation.level, "name")) %>%
  left_join(df_sample_stats, by = c(consolidation.level, "name"))

  df_combined <- df_combined %>% mutate(present = if_else(rpkm.mean > 0, 1, 0))
  df_combined$present <- factor(df_combined$present, levels = c(0, 1))
  
  df_combined <- df_combined %>%
    group_by(name) %>%
    mutate(zscore = as.numeric(scale(relabund.rpkm.mean.normalized))) %>%
    ungroup()
  
  df_combined <- df_combined %>%
    left_join(panel %>% select(Reporting_Name, Genus_Name, strand.type, taxid),
            by = c("name" = "Reporting_Name"))
  
  df_combined$name <- factor(df_combined$name, levels = df_combined %>%
  group_by(name) %>%
  summarize(total.norm.relabund = sum(relabund.rpkm.mean.normalized, na.rm = TRUE)) %>%
  arrange(desc(total.norm.relabund)) %>%
  pull(name))
  
  return(df_combined)
}

# only viruses present in df
# Summarize by location
df_influent_by.location_has.spike <- consolidate(df_influent_has_OC43_2, "WWTP") # spike intact
df_solids_by.location_has.spike <- consolidate(df_solids_has_OC43_2, "WWTP") # spike intact

df_influent_by.location_spike.removed <- consolidate(df_influent_spike.removed_2, "WWTP") # spike removed
df_solids_by.location_spike.removed <- consolidate(df_solids_spike.removed_2, "WWTP") # spike removed

# Summarize by fraction
df_vsp_samples_by.fraction_has.spike <- consolidate(df_single_spike_samples_has_OC43_2, "fraction") # spike intact
df_samples_by.fraction_spike.removed <- consolidate(df_samples_spike.removed_2, "fraction") # spike removed

# Summarize by fraction and locations
df_vsp_samples_by.WWTP.fraction_has.spike <- consolidate(df_single_spike_samples_has_OC43_2, "WWTP_fraction") # spike intact
df_samples_by.WWTP.fraction_spike.removed <- consolidate(df_samples_spike.removed_2, "WWTP_fraction") # spike removed


# direct compare - 5 WWTPs - without spike removed
df_samples_by.fraction_direct.compare_has.spike <- consolidate(df_single_spike_samples_has_OC43_direct.compare_2, "fraction")
df_single_by.WWTP.fraction_direct.compare_has.spike <- consolidate(df_single_spike_samples_has_OC43_direct.compare_2, "WWTP_fraction")
df_influent_by.location_direct.compare_has.spike <- consolidate(df_influent_has_OC43_direct.compare_2, "WWTP")
df_solids_by.location_direct.compare_has.spike <- consolidate(df_solids_has_OC43_direct.compare_2, "WWTP")

# direct compare - 5 WWTPs - spike removed
df_samples_by.fraction_direct.compare_spike.removed<- consolidate(df_samples_spike.removed_direct.compare_2, "fraction")
df_single_by.WWTP.fraction_direct.compare_spike.removed <- consolidate(df_samples_spike.removed_direct.compare_2, "WWTP_fraction")
df_influent_by.location_direct.compare_spike.removed <- consolidate(df_influent_spike.removed_direct.compare_2, "WWTP")
df_solids_by.location_direct.compare_spike.removed <- consolidate(df_solids_spike.removed_direct.compare_2, "WWTP")


# all viruses (undropped)
# Summarize by location
df_influent_by.location_has.spike_undropped <- consolidate(df_influent_has_OC43_2_undropped, "WWTP") # spike intact
df_solids_by.location_has.spike_undropped <- consolidate(df_solids_has_OC43_2_undropped, "WWTP") # spike intact

df_influent_by.location_spike.removed_undropped <- consolidate(df_influent_spike.removed_2_undropped, "WWTP") # spike removed
df_solids_by.location_spike.removed_undropped <- consolidate(df_solids_spike.removed_2_undropped, "WWTP") # spike removed

# Summarize by fraction
df_vsp_samples_by.fraction_has.spike_undropped <- consolidate(df_single_spike_samples_has_OC43_2_undropped, "fraction") # spike intact
df_samples_by.fraction_spike.removed_undropped <- consolidate(df_samples_spike.removed_2_undropped, "fraction") # spike removed

```

# Generate Top ~13 viruses dataframes (consolidated)
```{r}
top_viruses <- function(df, n){
  # Step 1: Find top ~13 names with highest abundance across all fractions
  top_names_df <- df %>%
    group_by(name) %>%
    summarize(sum_abundance = sum(relabund.rpkm.mean.normalized, na.rm = TRUE), .groups = "drop") %>%
    slice_max(order_by = sum_abundance, n = n, with_ties = FALSE)
  
  # Step 2: Filter original dataframe to just those top names
  df_top13 <- df %>%
    semi_join(top_names_df, by = "name")
  
  df_top13$name <- factor(df_top13$name, levels = top_names_df$name)
  return(df_top13)
}
df_top13_fraction_spike.removed <- top_viruses(df_samples_by.fraction_direct.compare_spike.removed, n=13)
df_top13_influent_spike.removed <- top_viruses(df_influent_by.location_direct.compare_spike.removed, n=13)
df_top13_solids_spike.removed <- top_viruses(df_solids_by.location_direct.compare_spike.removed, n=13)
```

# Heatmaps - Consolidated data
```{r}
library(taxize)
library(ape)
library(ggtree)
library(ggtreeExtra)
library(ggtext)
Sys.setenv(ENTREZ_KEY = "cf7f61606d77d41cbb9b23f88e1fa12f8908")

strand_colors <- c(
  "ssRNA"   = "#33BBEE",
  "dsRNA"   = "#0077BB",
  "ssDNA"   = "#FDB863",
  "dsDNA"   = "#EE7733",
  "pdsDNA"  = "#CC3311"
)
strand_order <- c("ssRNA", "dsRNA", "ssDNA", "dsDNA", "pdsDNA")
# Create the folder if it doesn't exist
if (!dir.exists("heatmaps")) dir.create("heatmaps")


heatmap.tree <- function(df, consolidation.level = c("fraction", "WWTP", "WWTP_fraction"), 
                         abundance_type = c("zscore", "fraction.present", "relabund.rpkm.mean.normalized", "log.relabund.norm"), 
                         width, height, mid, low, high, vsp.label.loc = -1.5, tree.prop = 0.65, strand.prop = 0.066){
  # df <- df_samples_by.WWTP.fraction_spike.removed
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  # Get unique taxid for each virus name
  unique.taxids <- df %>%
    distinct(name, taxid) %>%
    mutate(name = as.character(name))
  
  lineages <- classification(unique.taxids$taxid, db = "ncbi")
  
  # Build a tree from tax IDs
  tax_tree <- class2tree(lineages)$phylo
  
  # Replace tip labels on tree with my names list
  tax_tree$tip.label <- unique.taxids$name
  
  # Plot tree
  p_tree <- ggtree(tax_tree, layout = "rectangular")
  p_tree + geom_tiplab(size = 3, offset = 0.05, align = TRUE) +
    xlim(0, 200)  # adjust this upper limit to give more room for labels
  
  # 2) Extract the tip‐label order (top → bottom) from the tree
  tip_order <- p_tree$data %>%
    filter(isTip) %>% 
    arrange(y) %>%
    pull(label)
  
  # # Re‐factor your heatmap data to that order, and make BCoV spike grey
  df_reordered <- df %>%
    mutate(
      name = factor(name, levels = tip_order),  # order by tree tips
      name_label = if_else(                     # create a display label
        name == "Human coronavirus OC43 (HCoV_OC43)",
        "<span style='color:gray;'>Human coronavirus OC43 (HCoV_OC43)</span>",
        as.character(name)
     )
  )

  # Ensure strand.type is a factor and the associated virus names are ordered same as tree tips
  df_strand <- df_reordered %>%
    select(name, strand.type) %>%
    distinct() %>%
    mutate(
      name = factor(name, levels = tip_order),
      strand.type = factor(strand.type, levels = strand_order) # and the legend is in order
    )
  
  # Plot strand-type panel (one column)
  strand_panel <- ggplot(df_strand, aes(x = "Strand Type", y = name, fill = strand.type)) +
    geom_tile(color = "#696969") +
    scale_fill_manual(values = strand_colors, name = "Strand Type") +
    scale_y_discrete(position = "right") +
    theme_minimal() +
    theme(
      axis.title = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.text.y = element_blank(),
      panel.grid = element_blank(),
      plot.margin = margin(t = 25, r = 5, b = 5, l = 0)
    )
  
  heatmap.legend.title <- case_when(
    abundance_type == "zscore" ~ "Z-Score",
    abundance_type == "fraction.present" ~ "Fraction of Samples\nDetected in",
    abundance_type == "relabund.rpkm.mean.normalized" ~ "Relative Abundance",
    abundance_type == "log.relabund.norm" ~ "log₁₀(Relative Abundance)",
    TRUE ~ abundance_type  # fallback in case of an unexpected input
  )

  # Working heatmap
  heatmap_plot <- ggplot(df_reordered, aes(x = .data[[consolidation.level]], y = name, fill = .data[[abundance_type]])) +
    geom_tile(color = "#696969") +
    scale_fill_gradient2(
      low = "#0571b0", mid = "white", high = "#ca0020",
      midpoint = mid, limits = c(low, high), name = heatmap.legend.title,
      na.value = "#D6E2EE",
    ) +
    scale_y_discrete(position = "right") +
    theme_minimal() +
    theme(
      axis.text.x  = element_text(angle = 45, hjust = 1),
      axis.text.y = element_markdown(size = 8),
      axis.title   = element_blank(),
      panel.grid   = element_blank()
    )
  
  # Create grob for label
  vsp_label <- grid::textGrob(
    "VSP v2 Target",
    gp = gpar(fontsize = 10, fontface = "bold"),
    hjust = vsp.label.loc  # 2 = left, -2 = right
  )
  
  # Rebuild the heatmap with coord_cartesian(clip = "off") to prevent clipping
  heatmap_plot_labeled <- heatmap_plot +
    annotation_custom(
      grob = vsp_label,
      xmin = -Inf, xmax = Inf,
      ymin = length(levels(df_reordered$name)) + 1.5,  # move upward
      ymax = length(levels(df_reordered$name)) + 1.5
    ) +
    coord_cartesian(clip = "off")  # ⬅ prevents label clipping
  
  # Remove right margin of the tree and left margin of the heatmap
  p_tree <- p_tree + theme(plot.margin = margin(t = 5, r = 0, b = 5, l = 5))
  heatmap_plot_labeled <- heatmap_plot_labeled + theme(plot.margin = margin(t = 40, r = 60, b = 10, l = 0))
  
  # Combine the strand panel and z-score heatmap *first*, so their legends are aligned
  heatmap.panels <- strand_panel + heatmap_plot_labeled + 
    plot_layout(ncol = 2, widths = c(strand.prop, 1), guides = "collect") & 
    theme(legend.position = "right")
  
  # Combine the tree with the right panels
  final <- p_tree + heatmap.panels + plot_layout(ncol = 2, widths = c(tree.prop, 1))
  final
  
  print(final)
  
  # Save the plot to PDF
  ggsave(
    file     = file.path("heatmaps", paste0("heatmap_", abundance_type, "_", prefix, ".pdf")),
    plot     = final,                            # your final combined plot
    device   = cairo_pdf,                        # better text rendering
    width    = width,                               # adjust as needed
    height   = height,
    units    = "in"
  )
  
  # Return plot
  return(final)
}

heatmap.tree(df_samples_by.WWTP.fraction_spike.removed, "WWTP_fraction", "fraction.present", w=12, h=12, mid=0.0001, low=0.0001, high=1, vsp.label.loc = -1.85, tree.prop=0.65, strand.prop=0.066)
min(df_samples_by.WWTP.fraction_spike.removed$log.relabund.norm, na.rm = TRUE)
heatmap.tree(df_samples_by.WWTP.fraction_spike.removed, "WWTP_fraction", "log.relabund.norm", w=12, h=12, mid=-5.7, low=-5.7, high=0, vsp.label.loc = -1.75, tree.prop=0.65, strand.prop=0.066)

heatmap.tree(df_samples_by.fraction_spike.removed, "fraction", "fraction.present", w=8, h=12, mid=0.0001, low=0.0001, high=1, vsp.label.loc = -1.85, tree.prop=0.85, strand.prop=0.2)
min(df_samples_by.fraction_spike.removed$log.relabund.norm, na.rm = TRUE)
heatmap.tree(df_samples_by.fraction_spike.removed, "fraction", "log.relabund.norm", w=8.5, h=12, mid=-6.4, low=-6.4, high=0, vsp.label.loc = -1.75, tree.prop=0.85, strand.prop=0.2)

heatmap.tree(df_vsp_samples_by.fraction_has.spike_undropped, "fraction", "fraction.present", w=8, h=20, mid=0.0001, low=0.0001, high=1, vsp.label.loc = -1.85, tree.prop=0.85, strand.prop=0.2)
min(df_vsp_samples_by.fraction_has.spike_undropped$log.relabund.norm, na.rm = TRUE)
heatmap.tree(df_vsp_samples_by.fraction_has.spike_undropped, "fraction", "log.relabund.norm", w=8.5, h=20, mid=-7.6, low=-7.6, high=0, vsp.label.loc = -1.75, tree.prop=0.85, strand.prop=0.2)




```

# Strand type
```{r}
    
summarize.by.strand.type <- function(df, consolidation.level = c("fraction", "WWTP")) {
  consolidation.level <- match.arg(consolidation.level)

  # Perform summarization and assign to result_df
  result_df <- df %>%
    group_by(.data[[consolidation.level]], strand.type) %>%
    summarize(
      relabund.rpkm.mean.normalized = sum(relabund.rpkm.mean.normalized, na.rm = TRUE),
      .groups = "drop"
    ) %>%
    mutate(strand.type = factor(strand.type, levels = strand_order))

  return(result_df)
}

df_strand.type_by.fraction_spike.removed <- summarize.by.strand.type(df_samples_by.fraction_spike.removed, "fraction")
df_strand.type_by.fraction_spike.removed$fraction <- factor(df_strand.type_by.fraction_spike.removed$fraction,
                                                         levels = levels.fraction)

df_strand.type_influent_by.wwtp_spike.removed <- summarize.by.strand.type(df_influent_by.location_spike.removed, "WWTP")
df_strand.type_influent_by.wwtp_spike.removed$WWTP <- factor(df_strand.type_influent_by.wwtp_spike.removed$WWTP,
                                                         levels = levels.WWTP)

df_strand.type_solids_by.wwtp_spike.removed <- summarize.by.strand.type(df_solids_by.location_spike.removed, "WWTP")
df_strand.type_solids_by.wwtp_spike.removed$WWTP <- factor(df_strand.type_solids_by.wwtp_spike.removed$WWTP,
                                                         levels = levels.WWTP)


df_strand.type_by.fraction_direct.compare_spike.removed <- summarize.by.strand.type(df_samples_by.fraction_direct.compare_spike.removed, "fraction")
df_strand.type_by.fraction_direct.compare_spike.removed$fraction <- factor(df_strand.type_by.fraction_direct.compare_spike.removed$fraction,
                                                         levels = levels.fraction)

df_strand.type_influent_by.wwtp_direct.compare_spike.removed <- summarize.by.strand.type(df_influent_by.location_direct.compare_spike.removed, "WWTP")
df_strand.type_influent_by.wwtp_direct.compare_spike.removed$fraction <- factor(df_strand.type_influent_by.wwtp_direct.compare_spike.removed$WWTP,
                                                         levels = levels.WWTP)

df_strand.type_solids_by.wwtp_direct.compare_spike.removed <- summarize.by.strand.type(df_solids_by.location_direct.compare_spike.removed, "WWTP")
df_strand.type_solids_by.wwtp_direct.compare_spike.removed$fraction <- factor(df_strand.type_solids_by.wwtp_direct.compare_spike.removed$WWTP,
                                                         levels = levels.WWTP)


plot.strand.type <- function(df, consolidation.level = c("fraction", "WWTP")) {
  consolidation.level <- match.arg(consolidation.level)

  p_strand <- ggplot(df, 
                     aes(x = .data[[consolidation.level]], y = relabund.rpkm.mean.normalized, fill = strand.type)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = strand_colors) +
  labs(
    x = NULL,
    y = "Normalized Relative Abundance (RPKM)",
    fill = "Strand Type"
  ) +
  theme_minimal()

p_strand
}
plot.strand.type(df_strand.type_by.fraction_spike.removed, "fraction")
plot.strand.type(df_strand.type_influent_by.wwtp_spike.removed, "WWTP")
plot.strand.type(df_strand.type_solids_by.wwtp_spike.removed, "WWTP")

plot.strand.type(df_strand.type_by.fraction_direct.compare_spike.removed, "fraction")
plot.strand.type(df_strand.type_influent_by.wwtp_direct.compare_spike.removed, "WWTP")
plot.strand.type(df_strand.type_solids_by.wwtp_direct.compare_spike.removed, "WWTP")

```

# Per-virus plots
```{r}

consolidated.plot.by.virus <- function(df, consolidation.level = c("fraction", "WWTP"), width, height)  {
  consolidation.level <- match.arg(consolidation.level)
  # Extract the prefix from the input df name
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  
  sample_fraction <- case_when(
    grepl("solids", prefix)   ~ "in Solids",
    grepl("influent", prefix) ~ "in Influent",
    TRUE                      ~ "")
  
  title_base <- paste("Viruses", sample_fraction, "Consolidated to", tools::toTitleCase(consolidation.level))
  ref_df <- paste0("(", prefix, ")")
  
  # Individual titles
  title <- paste("Relative Abundance of", title_base, "\n", ref_df)

  p <- ggplot(df, aes(x = name, y = relabund.rpkm.mean.normalized, fill = .data[[consolidation.level]])) +
    geom_bar(stat = "identity", color = "Black") +
    scale_x_discrete(expand = c(0.05, 0.09)) +
    labs(
      title = title,
      x = "Virus",
      y = "Mean Relative Abundance (RPKM)",
      fill = tools::toTitleCase(consolidation.level)
    ) +
    theme_minimal() +
    scale_fill_manual(values = custom.colors) +
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.grid = element_blank(),
      axis.ticks.x = element_line(),
      axis.ticks.y = element_line(),
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.margin = margin(10, 10, 10, 50)
    )

  # Save plots in separate folders
  ggsave(file = file.path("by_virus_consolidated", paste0("plot_viruses_relabund_", consolidation.level, "_", prefix, ".pdf")), 
         plot = p, width = width, height = height)

  # Return plot
  return(p)
}

if (!dir.exists("by_virus_consolidated")) {
  dir.create("by_virus_consolidated")
}

consolidated.plot.by.virus(df_influent_by.location_has.spike, "WWTP", w=15, h=10)
consolidated.plot.by.virus(df_solids_by.location_has.spike, "WWTP", w=15, h=10)

consolidated.plot.by.virus(df_influent_by.location_spike.removed, "WWTP", w=15, h=10)
consolidated.plot.by.virus(df_solids_by.location_spike.removed, "WWTP", w=15, h=10)

# By fraction
consolidated.plot.by.virus(df_vsp_samples_by.fraction_has.spike, "fraction", w=15, h=10)
consolidated.plot.by.virus(df_samples_by.fraction_spike.removed, "fraction", w=15, h=10)


# direct compare - 5 WWTPs - without spike removed
consolidated.plot.by.virus(df_samples_by.fraction_direct.compare_has.spike, "fraction", w=15, h=10)
consolidated.plot.by.virus(df_influent_by.location_direct.compare_has.spike, "WWTP", w=15, h=10)
consolidated.plot.by.virus(df_solids_by.location_direct.compare_has.spike, "WWTP", w=15, h=10)

# direct compare - 5 WWTPs - spike removed
consolidated.plot.by.virus(df_samples_by.fraction_direct.compare_spike.removed, "fraction", w=15, h=10)
consolidated.plot.by.virus(df_influent_by.location_direct.compare_spike.removed, "WWTP", w=15, h=10)
consolidated.plot.by.virus(df_solids_by.location_direct.compare_spike.removed, "WWTP", w=15, h=10)

# Top 13 plots
consolidated.plot.by.virus(df_top13_fraction_spike.removed, "fraction", w=10, h=7)
consolidated.plot.by.virus(df_top13_influent_spike.removed, "WWTP", w=10, h=7)
consolidated.plot.by.virus(df_top13_solids_spike.removed, "WWTP", w=10, h=7)

```

# UpSet Plots
```{r}
upset.plot <- function(df, consolidation.level = c("fraction", "WWTP", "WWTP_fraction"), width, height, top, bottom)  {
  consolidation.level <- match.arg(consolidation.level)
  # Extract the prefix from the input df name
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  
  p_bar <- ggplot(df, aes(x = name, y = relabund.rpkm.mean, fill = .data[[consolidation.level]])) +
  geom_bar(stat = "identity", color = "black") +
  scale_fill_manual(values = custom.colors) +
  labs(
    x = NULL,
    y = "Relative Abundance (RPKM)"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_blank(), 
    axis.ticks.x = element_blank(),
    panel.grid = element_blank(),
    plot.margin = margin(0, 0, 0, 0)
  )

  # Make a copy of the df with reverse order factored levels for WWTP or fraction to correspond with bar plot above
  
  df_matrix <- df %>%
  mutate(!!consolidation.level := forcats::fct_rev(.data[[consolidation.level]]))
  
  p_matrix <- ggplot(df_matrix, aes(x = name, y = .data[[consolidation.level]], fill = .data[[consolidation.level]], alpha = present)) +
  geom_tile(color = "#696969") +
  scale_alpha_manual(values = c("0" = 0, "1" = 1)) +
  scale_fill_manual(values = custom.colors) +
  labs(
    x = "Virus",
    y = NULL,
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, color = "black"),
    axis.text.y = element_text(color = "black"),
    legend.position = "none",
    panel.grid.major = element_blank(),
    plot.margin = margin(0, 0, 0, 0)
  )
  
  # p <- p_bar / p_matrix + plot_layout(heights = c(3, 0.75))
  # Stack without extra gap
  p <- p_bar / p_matrix + plot_layout(heights = c(top, bottom))
  
  # Add outer margin only
  p <- p & plot_annotation(theme = theme(plot.margin = margin(10, 10, 10, 80)))
  
  # Save plots in separate folders
  ggsave(file = file.path("by_virus_upset", paste0("plot_viruses_relabund_", consolidation.level, "_", prefix, ".pdf")), 
         plot = p, width = width, height = height)

  # Return plot
  return(p)
}

if (!dir.exists("by_virus_upset")) {
  dir.create("by_virus_upset")
}

upset.plot(df_influent_by.location_has.spike, "WWTP", w=15, h=10, t=2, b=0.5)
upset.plot(df_solids_by.location_has.spike, "WWTP", w=15, h=10, t=2, b=0.5)

upset.plot(df_influent_by.location_spike.removed, "WWTP", w=15, h=10, t=2, b=0.5)
upset.plot(df_solids_by.location_spike.removed, "WWTP", w=15, h=10, t=2, b=0.5)

# By fraction
upset.plot(df_vsp_samples_by.fraction_has.spike, "fraction", w=18, h=10, t=2, b=0.17)
upset.plot(df_samples_by.fraction_spike.removed, "fraction", w=18, h=10, t=2, b=0.17)

# By WWTP and fraction
upset.plot(df_vsp_samples_by.WWTP.fraction_has.spike, "WWTP_fraction", w=22, h=10, t=2, b=0.6)
upset.plot(df_samples_by.WWTP.fraction_spike.removed, "WWTP_fraction", w=22, h=10, t=2, b=0.6)


```

```{r}
# Healthy Places Indexes
hpi <- read_excel("../../inputs-sequoia/healthy-places.xlsx")
```

# Non-consolidated abundance - controls, weird samples
```{r}
plot_abundance <- function(df, width, height, ncol)  {
  # Extract the prefix from the input df name
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  
  sample_fraction <- case_when(
    grepl("controls", prefix)       ~ "Controls",
    grepl("nonsamples", prefix)     ~ "Non-Samples",
    grepl("sample\\.spike", prefix) ~ "Tri-Spike Replicates",
    TRUE                            ~ "")
  
  title_base <- paste("of Viruses in", sample_fraction)
  ref_df <- paste0("(", prefix, ")")
  
  # Individual titles
  relabund_title <- paste("Relative Abundance of", title_base, "\n", ref_df)
  abundance_title <- paste("Raw Abundance of", title_base, "\n", ref_df)
  relabund_label <- "Relative Abundance"
  abundance_label <- "Raw Abundance"
  plot_title <- paste("Relative and Raw Abundance of", title_base, "\n", ref_df)

  # Base plot styling
  base_theme <- theme_minimal() +
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.grid = element_blank(),
      axis.ticks.x = element_line(),
      axis.ticks.y = element_line(),
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.margin = margin(10, 10, 10, 10)
    )
  
# Plot 1: Relative Abundance (relabund.rpkm)
  p_relabund <- ggplot(df, aes(x = sample, y = relabund.rpkm, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(limits = c(0, 1.01), expand = c(0, 0)) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      title = relabund_title,
      x = sample_fraction,
      y = "Relative Abundance of RPKM",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme

  # Plot 2: Raw RPKM abundance
  p_rpkm <- ggplot(df, aes(x = sample, y = rpkm, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      title = abundance_title,
      x = sample_fraction,
      y = "RPKM",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme
  
  # Combining relabund and abundance plots
  # Remove legend from left plot
  p_relabund_no_legend <- p_relabund + theme(legend.position = "none") + labs(title = relabund_label)
  p_rpkm_with_legend <- p_rpkm + labs(title = abundance_label)

  p_combined <- (p_relabund_no_legend | p_rpkm_with_legend) +
  plot_layout(guides = "collect") +
    plot_annotation(
    title = plot_title,
    theme = theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5, margin = margin(b = 10)))
  ) &
  theme(legend.position = "right")
  
  # Save plots in separate folders
  ggsave(file = file.path("nonconsolidated_graphs/1.relabund", paste0("plot_relabund_", prefix, ".pdf")), 
         plot = p_relabund, width = width, height = height)
  
  ggsave(file = file.path("nonconsolidated_graphs/2.rpkm", paste0("plot_rpkm_", prefix, ".pdf")), 
         plot = p_rpkm, width = width, height = height)
  
  ggsave(file = file.path("nonconsolidated_graphs/3.combined", paste0("plot_combined_", prefix, ".pdf")), 
         plot = p_combined, width = width * 1.5, height = height)

  # Return both plots as a named list
  return(list(
    relabund_plot = p_relabund,
    rpkm_plot = p_rpkm,
    combined_plot = p_combined
  ))
}

# Create directories if they don't already exist
dir.create("nonconsolidated_graphs/1.relabund", showWarnings = FALSE, recursive = TRUE)
dir.create("nonconsolidated_graphs/2.rpkm", showWarnings = FALSE, recursive = TRUE)
dir.create("nonconsolidated_graphs/3.combined", showWarnings = FALSE, recursive = TRUE)

plots_vsp_controls <- plot_abundance(df_vsp_controls_processed, w=8, h=6.5, n=1)
plots_vsp_nonsamples <- plot_abundance(df_vsp_nonsamples_processed, w=17, h=6.5, n=3)
plots_vsp_tri.spike.samples_tri.spike.removed <- plot_abundance(df_tri_spike_samples_spike.removed_processed, w=8, h=7, n=1)
plots_vsp_tri.spike.samples <- plot_abundance(df_vsp_tri.spike.samples_processed, w=10, h=7, n=1)


# # View them
# plots_vsp_controls$relabund_plot
# plots_vsp_controls$rpkm_plot
# plots_vsp_controls$combined_plot

```

# Function to plot average consolidated abundance
```{r}
plot_consolidated_abundance <- function(df, consolidation.level = c("fraction", "WWTP"), width, height, ncol)  {
  consolidation.level <- match.arg(consolidation.level)
  # Extract the prefix from the input df name
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  
  sample_fraction <- case_when(
    grepl("solids", prefix)   ~ "in Solids",
    grepl("influent", prefix) ~ "in Influent",
    TRUE                      ~ "")
  
  title_base <- paste("Viruses", sample_fraction, "Consolidated to", tools::toTitleCase(consolidation.level))
  ref_df <- paste0("(", prefix, ")")
  
  # Individual titles
  relabund_title <- paste("Relative Abundance of", title_base, "\n", ref_df)
  abundance_title <- paste("Raw Abundance of", title_base, "\n", ref_df)
  relabund_label <- "Relative Abundance"
  abundance_label <- "Raw Abundance"
  plot_title <- paste("Relative and Raw Abundance of", title_base, "\n", ref_df)

  # Base plot styling
  base_theme <- theme_minimal() +
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.grid = element_blank(),
      axis.ticks.x = element_line(),
      axis.ticks.y = element_line(),
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.margin = margin(10, 10, 10, 10)
    )
  
# Plot 1: Relative Abundance (relabund.rpkm)
  p_relabund <- ggplot(df, aes(x = .data[[consolidation.level]], y = relabund.rpkm.mean.normalized, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(limits = c(0, 1.01), expand = c(0, 0)) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      title = relabund_title,
      x = tools::toTitleCase(consolidation.level),
      y = "Relative Abundance of RPKM",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme

  # Plot 2: Raw RPKM abundance
  p_rpkm <- ggplot(df, aes(x = .data[[consolidation.level]], y = rpkm.mean, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      title = abundance_title,
      x = tools::toTitleCase(consolidation.level),
      y = "Average RPKM",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme
  
  # Combining relabund and abundance plots
  # Remove legend from left plot
  p_relabund_no_legend <- p_relabund + theme(legend.position = "none") + labs(title = relabund_label)
  p_rpkm_with_legend <- p_rpkm + labs(title = abundance_label)

  p_combined <- (p_relabund_no_legend | p_rpkm_with_legend) +
  plot_layout(guides = "collect") +
    plot_annotation(
    title = plot_title,
    theme = theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5, margin = margin(b = 10)))
  ) &
  theme(legend.position = "right")
  
  # Save plots in separate folders
  ggsave(file = file.path("consolidated_graphs/1.relabund", paste0("plot_relabund_", consolidation.level, "_", prefix, ".pdf")), 
         plot = p_relabund, width = width, height = height)
  
  ggsave(file = file.path("consolidated_graphs/2.rpkm", paste0("plot_rpkm_", consolidation.level, "_", prefix, ".pdf")), 
         plot = p_rpkm, width = width, height = height)
  
  ggsave(file = file.path("consolidated_graphs/3.combined", paste0("plot_combined_", consolidation.level, "_", prefix, ".pdf")), 
         plot = p_combined, width = width * 1.5, height = height)

  # Return both plots as a named list
  return(list(
    relabund_plot = p_relabund,
    rpkm_plot = p_rpkm,
    combined_plot = p_combined
  ))
}

```

# Experimental: consolidated normalized plots
```{r}
plot_consolidated_abundance <- function(df, consolidation.level = c("fraction", "WWTP"), width, height, ncol) {
  consolidation.level <- match.arg(consolidation.level)
  
  # Extract the prefix from the input df name
  prefix <- sub("_processed$", "", sub("^df_", "", deparse(substitute(df))))
  
  sample_fraction <- case_when(
    grepl("solids", prefix)   ~ "in Solids",
    grepl("influent", prefix) ~ "in Influent",
    TRUE                      ~ ""
  )
  
  title_base <- paste("Viruses", sample_fraction, "Consolidated to", tools::toTitleCase(consolidation.level))
  ref_df <- paste0("(", prefix, ")")
  
  # Titles
  plot_title <- paste("Relative and Raw Abundance of", title_base, "\n", ref_df)
  
  rpkm_label <- "Total Abundance (RPKM)"
  recovery_label <- "RPKM normalized to BCoV rec."
  rpkm_norm_label <- "RPKM normalized to BCoV RPKM"
  
  # Base plot styling
  base_theme <- theme_minimal() +
    theme(
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      panel.grid = element_blank(),
      axis.ticks.x = element_line(),
      axis.ticks.y = element_line(),
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      plot.margin = margin(10, 10, 10, 10)
    )
  
  # Plot 1: Raw RPKM abundance
  p_rpkm <- ggplot(df, aes(x = .data[[consolidation.level]], y = rpkm.mean, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      x = tools::toTitleCase(consolidation.level),
      y = "RPKM Averaged to Location",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme
  
  # Plot 2: Mean RPKM normalized to BCoV recovery
  p_recovery <- ggplot(df, aes(x = .data[[consolidation.level]], y = mean.rpkm.norm.to.BCoV.recovery, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      x = tools::toTitleCase(consolidation.level),
      y = "Mean of RPKM/(BCoV Recovery)",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme
  
  # Plot 3: Mean RPKM normalized to BCoV RPKM
  p_rpkm_norm <- ggplot(df, aes(x = .data[[consolidation.level]], y = mean.rpkm.norm.to.BCoV.rpkm, fill = name)) +
    geom_bar(stat = "identity", color = "Dark Blue") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_x_discrete(expand = c(0.09, 0.09)) +
    labs(
      x = tools::toTitleCase(consolidation.level),
      y = "Mean of RPKM/(BCoV RPKM)",
      fill = "Viruses"
    ) +
    guides(fill = guide_legend(ncol = ncol)) +
    base_theme
  
  p_recovery_no_legend <- p_recovery + theme(legend.position = "none")
  p_rpkm_norm_no_legend <- p_rpkm_norm + theme(legend.position = "none")
  p_rpkm_with_legend <- p_rpkm

  
  p_combined <- (p_rpkm | p_recovery_no_legend | p_rpkm_norm_no_legend) +
  plot_layout(guides = "collect") +
    plot_annotation(
    title = plot_title,
    theme = theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5, margin = margin(b = 10)))
  ) &
  theme(legend.position = "right")
  
  
  # Save the plots
  ggsave(file = file.path("consolidated_graphs/1.raw.rpkm", paste0("plot_raw_rpkm_", consolidation.level, "_", prefix, ".pdf")),
         plot = p_rpkm, width = width, height = height)
  
  ggsave(file = file.path("consolidated_graphs/2.rpkm.norm.to.recovery", paste0("plot_norm_to_recovery_", consolidation.level, "_", prefix, ".pdf")),
         plot = p_recovery, width = width, height = height)
  
  ggsave(file = file.path("consolidated_graphs/3.rpkm.norm.to.rpkm", paste0("plot_norm_to_rpkm_", consolidation.level, "_", prefix, ".pdf")),
         plot = p_rpkm_norm, width = width, height = height)
  
  ggsave(file = file.path("consolidated_graphs/4.combined", paste0("plot_combined_", consolidation.level, "_", prefix, ".pdf")),
       plot = p_combined, width = width * 1.5, height = height)
  
  # Return the plots as a named list
  return(list(
    rpkm_plot = p_rpkm,
    recovery_plot = p_recovery,
    rpkm_norm_plot = p_rpkm_norm,
    combined_plot = p_combined
  ))
}
```

# Make consolidated plots
```{r}
# Create directories if they don't already exist
dir.create("consolidated_graphs/1.relabund", showWarnings = FALSE, recursive = TRUE)
dir.create("consolidated_graphs/2.rpkm", showWarnings = FALSE, recursive = TRUE)
dir.create("consolidated_graphs/3.combined", showWarnings = FALSE, recursive = TRUE)

# Plot by location
plots_vsp_influent_by_location <- plot_consolidated_abundance(df_influent_by.location_has.spike, 
                                              "WWTP", w=15, h=10, n=2) # spike intact
plots_vsp_influent_bcov.spike.removed_by_location <- plot_consolidated_abundance(df_influent_by.location_spike.removed2, 
                                                                 "WWTP", w=15, h=10, n=2) # spike removed
plots_vsp_solids_by_location <- plot_consolidated_abundance(df_solids_by.location_has.spike, 
                                            "WWTP", w=15, h=10, n=2) # spike intact
plots_vsp_solids_bcov.spike.removed_by_location <- plot_consolidated_abundance(df_solids_by.location_spike.removed2, 
                                                               "WWTP", w=15, h=10, n=2) # spike removed

# Plot by fraction
plots_vsp_samples_by_fraction <- plot_consolidated_abundance(df_vsp_samples_by.fraction_has.spike, 
                                         "fraction", w=15, h=10, n=3) # spike intact

plots_vsp_samples_bcov.spike.removed_by_fraction <- plot_consolidated_abundance(df_samples_by.fraction_spike.removed2, 
                                                            "fraction", w=15, h=10, n=3) # spike removed




# # View them
# plots_vsp_influent_by_location$relabund_plot
# plots_vsp_influent_by_location$rpkm_plot
# plots_vsp_influent_by_location$combined_plot

```


# Plot just BCoV and Adenovirus F to find outliers
```{r}
scatter.single.viruses <- function(df, virus, type){
  df_filtered <- df %>%
    filter(name == virus)
    
  p_relbund <- ggplot(df_filtered, aes(x = sample, y = relabund.rpkm)) +
    geom_point() +
    scale_y_continuous(limits = c(0, 1)) +
    labs(x = NULL, y = "Relative Abundance (RPKM)") +
    theme_minimal() +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank()
    )
  
  p_rpkm <- ggplot(df_filtered, aes(x = sample, y = rpkm)) +
    geom_point() +
    labs(x = "Sample", y = "Abundance (RPKM)") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  p <- (p_relbund / p_rpkm) + patchwork::plot_annotation(title = paste(virus, "-", type))
  
  return(p)
}

scatter.single.viruses(df_solids_spike.removed_2, virus="Human adenovirus F", type="Solids")
scatter.single.viruses(df_solids_has_OC43_2, virus=bcov.spike, type="Solids")

scatter.single.viruses(df_influent_spike.removed_2, virus="Human adenovirus F", type="Influent")
scatter.single.viruses(df_influent_has_OC43_2, virus=bcov.spike, type="Influent")

```

# SARS-CoV-2 plots
```{r}
sars <- "Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)"

df_sars_influent <- df_influent_spike.removed_2_undropped %>% filter(name == sars)
df_sars_solids <- df_solids_spike.removed_2_undropped %>% filter(name == sars)

df_sars_influent_2 <- df_sars_influent %>%
  left_join(
    influent_ddpcr %>% select(sample.edit, `N1_Concentration_Merged`, `N2_Concentration_Merged`),
    by = "sample.edit"
  )

df_sars_solids_2 <- df_sars_solids %>%
  left_join(
    solids_ddpcr %>% select(sample.edit, `N_Gene_gc_per_g_dry_weight`, `S_Gene_gc_per_g_dry_weight`),
    by = "sample.edit"
  ) %>%
  mutate(
    N_Gene_gc_per_g_dry_weight = as.numeric(N_Gene_gc_per_g_dry_weight),
    S_Gene_gc_per_g_dry_weight = as.numeric(S_Gene_gc_per_g_dry_weight)
  )

library(lubridate)

df_sars_influent_2 <- df_sars_influent_2 %>%
  mutate(
    date_str = paste0("2022", substr(sample.edit, 1, 4)),  # Add year prefix
    date = ymd(date_str)  # Convert to Date
  )

df_sars_solids_2 <- df_sars_solids_2 %>%
  mutate(
    date_str = paste0("2022", substr(sample.edit, 1, 4)),  # Add year prefix
    date = ymd(date_str)  # Convert to Date
  )

dir.create("sars-cov-2", showWarnings = FALSE)
dir.create("sars-cov-2/sars-inf", showWarnings = FALSE)
dir.create("sars-cov-2/sars-sol", showWarnings = FALSE)
dir.create("sars-cov-2/sars-combined", showWarnings = FALSE)

########################################
sars.by.date <- function(city.name){
  
  df_inf <- df_sars_influent_2 %>%
    filter(city == city.name)
  
  df_sol <- df_sars_solids_2 %>%
    filter(city == city.name)
  
  # Define base theme
  base_theme <- theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  # Define base scales
  base_scales <- list(
    scale_color_manual(values = c("N1" = "lightskyblue1", "N2" = "steelblue1", "N Gene" = "mediumblue", "S Gene" = "navy", "Relative\nAbundance\n(RPKM)" = "red")),
    scale_x_date(
      limits = as.Date(c("2022-05-01", "2022-12-31")),
      date_breaks = "1 month",
      date_labels = "%b"
    )
  )
  
  # ddPCR conc plot - Influent
  p_sars_conc_inf <- ggplot(df_inf, aes(x = date)) +
    geom_line(aes(y = N1_Concentration_Merged, color = "N1")) +
    geom_point(aes(y = N1_Concentration_Merged, color = "N1")) +
    geom_line(aes(y = N2_Concentration_Merged, color = "N2")) +
    geom_point(aes(y = N2_Concentration_Merged, color = "N2")) +
    labs(x=NULL, y = "Concentration (copies/mL)", color = "ddPCR Probe\n(UCD)") +
    base_scales +
    base_theme +
    scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0.005, 0.05)))
  p_sars_conc_inf
  
  # Relative abundance plot - Influent
  p_sars_relabund_inf <- ggplot(df_inf, aes(x = date)) +
    geom_line(aes(y = relabund.rpkm, color = "Relative\nAbundance\n(RPKM)")) +
    geom_point(aes(y = relabund.rpkm, color = "Relative\nAbundance\n(RPKM)")) +
    labs(x = NULL, y = "Relative Abundance of SARS-CoV-2 (RPKM)", color = NULL) +
    base_scales +
    base_theme +
    scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0.005, 0.05)))
  p_sars_relabund_inf
  
  p_sars_inf <- (p_sars_conc_inf / p_sars_relabund_inf) + patchwork::plot_annotation(title = paste(city.name, "-", "Influent"))
  p_sars_inf
  ggsave(file = file.path("sars-cov-2/sars-inf", paste0(city.name, "_", "influent", ".pdf")), 
         plot = p_sars_inf, width = 7.4, height = 8)
  
  
  p_sars_conc_sol <- ggplot() +  # no global data here anymore
    # N Gene
    geom_line(
      data = df_sol %>% filter(!is.na(N_Gene_gc_per_g_dry_weight)),
      aes(x = date, y = as.numeric(N_Gene_gc_per_g_dry_weight), color = "N Gene", group = "N Gene")
    ) +
    geom_point(
      data = df_sol %>% filter(!is.na(N_Gene_gc_per_g_dry_weight)),
      aes(x = date, y = as.numeric(N_Gene_gc_per_g_dry_weight), color = "N Gene", group = "N Gene")
    ) +
    geom_line(
      data = df_sol %>% filter(!is.na(S_Gene_gc_per_g_dry_weight)),
      aes(x = date, y = as.numeric(S_Gene_gc_per_g_dry_weight), color = "S Gene", group = "S Gene")
    ) +
    geom_point(
      data = df_sol %>% filter(!is.na(S_Gene_gc_per_g_dry_weight)),
      aes(x = date, y = as.numeric(S_Gene_gc_per_g_dry_weight), color = "S Gene", group = "S Gene")
    ) +
    labs(
      x = NULL,
      y = "Concentration (copies/g dry weight)",
      color = "ddPCR Probe\n(Eurofins)"
    ) +
    base_scales +
    base_theme +
    scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0.005, 0.05)))
  p_sars_conc_sol
  
  p_sars_relabund_sol <- ggplot(df_sol, aes(x = date)) +
    geom_line(aes(y = relabund.rpkm, color = "Relative\nAbundance\n(RPKM)")) +
    geom_point(aes(y = relabund.rpkm, color = "Relative\nAbundance\n(RPKM)")) +
    labs(x=NULL, y = "Relative Abundance of SARS-CoV-2 (RPKM)", color = NULL) +
    base_scales +
    base_theme +
    scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0.005, 0.05)))
  p_sars_relabund_sol
  
  p_sars_sol <- (p_sars_conc_sol / p_sars_relabund_sol) + 
    patchwork::plot_annotation(title = paste(city.name, "-", "Solids"))
  p_sars_sol
  ggsave(file = file.path("sars-cov-2/sars-sol", paste0(city.name, "_", "solids", ".pdf")), 
         plot = p_sars_sol, width = 7.4, height = 8)
  
  # Get legend info
  library(cowplot)
  legend_conc_inf <- get_legend(p_sars_conc_inf)
  legend_conc_sol <- get_legend(p_sars_conc_sol)
  legend_relabund <- get_legend(p_sars_relabund_inf)
  
  # Take off all legends
  p_sars_conc_inf <- p_sars_conc_inf + theme(legend.position = "none")
  p_sars_relabund_inf <- p_sars_relabund_inf + theme(legend.position = "none")
  p_sars_conc_sol <- p_sars_conc_sol + theme(legend.position = "none")
  p_sars_relabund_sol <- p_sars_relabund_sol + theme(legend.position = "none")
  
  p_sars_inf_2 <- (p_sars_conc_inf / p_sars_relabund_inf) + labs(title = "Influent") & 
  theme(plot.title = element_text(hjust = 0.5))
  p_sars_sol_2 <- (p_sars_conc_sol / p_sars_relabund_sol) + labs(title = "Solids") & 
  theme(plot.title = element_text(hjust = 0.5))
  p_sars_inf_2
  p_sars_sol_2
  
  
  legend_stack <- patchwork::wrap_elements(full = legend_conc_inf) /
                  patchwork::wrap_elements(full = legend_conc_sol) /
                  patchwork::wrap_elements(full = legend_relabund)
  
  p_sars <- (p_sars_inf_2 | p_sars_sol_2 | legend_stack) + plot_layout(widths = c(1, 1, 0.3)) + 
    patchwork::plot_annotation(title = paste(city.name, "- SARS-CoV-2"))
  p_sars
  
  # Save plots in separate folders
  ggsave(file = file.path("sars-cov-2/sars-combined", paste0(city.name, "_", "combined_SARS", ".pdf")), 
         plot = p_sars, width = 12, height = 8)
  
  return(p_sars)
}
sars.by.date(city.name="Davis")
sars.by.date(city.name="Esparto")
sars.by.date(city.name="Los Banos")
sars.by.date(city.name="Turlock")
sars.by.date(city.name="Winters")
sars.by.date(city.name="Woodland")
sars.by.date(city.name="Merced")
sars.by.date(city.name="Modesto")


```

# genus level analysis
```{r}
process_json <- function(json_files) {
  df_all <- data.frame()
  
  for (file in json_files) {
    # Load JSON
    json_data <- fromJSON(file)
      
    df <- data.frame(
    Sample = sub("\\..*", "", basename(file)),
    Genus = names(json_data),
    Abundance = as.numeric(json_data),
    row.names = NULL,
    stringsAsFactors = FALSE
    )
  
    df$RelAbun <- (df$Abundance / sum(df$Abundance))
    
    # Append to main dataframe
    df_all <- rbind(df_all, df)
    }
  
  df_all$fraction <- case_when(
    grepl("^Inh", df_all$Sample) ~ "Influent",
    grepl("^Euro", df_all$Sample) ~ "Solids",
    TRUE ~ "Other"
  )
  
  df_all %>%
  group_by(Genus) %>%
  mutate(TotalAbundance = sum(RelAbun)) %>%
  ungroup()

  df_all$fraction <- factor(df_all$fraction, levels = c("Solids", "Influent"))

  df_all$Genus <- factor(df_all$Genus, levels = df_all %>%
                         group_by(Genus) %>%
                         summarize(TotalAbundance = sum(RelAbun)) %>%
                         arrange(TotalAbundance) %>%
                         pull(Genus))

  df_all$Sample <- gsub("^Euro-|^Inh-", "", df_all$Sample)
  
  return(df_all)
}

df_amit_v1 <- process_json(json_files_amit_v1)
df_amit_v2 <- process_json(json_files_amit_v2)
df_dragen_v1 <- process_json(json_files_dragen_v1)
df_dragen_v2 <- process_json(json_files_dragen_v2)

```


#ggplot genuses
```{r}


p_genus <- ggplot(df_amit_v2, aes(x = Sample, y = RelAbun, fill = Genus)) +
  geom_bar(stat = "identity", color = "Dark Blue") +
  theme_minimal() +
  scale_y_continuous(limits = c(0, 1.01), expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  labs(x = "Sample",
       y = "Relative Abundance (kraken VSP v1)",
       fill = "Virus Genus") +
  theme(
    panel.grid = element_blank(),    # Removes all gridlines (major & minor)
    axis.ticks.x = element_line(),   # Keeps x-axis ticks
    axis.ticks.y = element_blank(),  # Removes y-axis ticks
    axis.text.x = element_text(angle = 45, hjust = 1),# Keeps x-axis labels readable
    panel.spacing = unit(1, "cm"),   # Increases space between facet panels
    legend.position = "right",
    # Adjust legend font size
    legend.text = element_text(size = 7),  # Change legend text size
    legend.title = element_text(size = 10), # Change legend title size
    legend.key.size = unit(0.4, "cm")  # Reduce legend key box size
  ) +
  facet_wrap(~ fraction, scales = "free_x") +
  guides(fill = guide_legend(ncol = 7))

pdf("p_genus.pdf", width = 20, height = 15)
print(p_genus)
dev.off()
p_genus
```
